src/map.c

```diff
@@ -1,754 +1,592 @@
+                rbtn_black_set(left);
+                rbtn_black_set(right);
+                rbtn_red_set(cnode);
+            } else {
+                /* Lean left. */
+                map_node_t *tnode;
+                bool tred = rbtn_red_get(cnode);
+                rbtn_rotate_left(map_node_t, link, cnode, tnode);
+                rbtn_color_set(tnode, tred);
+                rbtn_red_set(cnode);
+                cnode = tnode;
+            }
+        }
+        pathp->node = cnode;
+    }
+    /* Set root, and make it black. */
+    rbtree->root = path->node;
+    rbtn_black_set(rbtree->root);
 }
 
-static void map_fix_colors(map_t obj, map_node_t *node)
+static void internal_map_remove(map_internal_t *rbtree, map_node_t *node)
 {
-    /* If root, set the color to black */
-    if (node == obj->head) {
-        rb_set_color(node, RB_BLACK);
-        return;
+    internal_map_path_entry_t path[RB_MAX_DEPTH];
+    internal_map_path_entry_t *pathp = NULL;
+    internal_map_path_entry_t *nodep = NULL;
+    /* Wind. */
+    path->node = rbtree->root;
+    for (pathp = path; pathp->node; pathp++) {
+        int cmp = pathp->cmp = (rbtree->cmp)(node, pathp->node);
+        if (cmp < 0) {
+            pathp[1].node = rbtn_left_get(pathp->node);
+        } else {
+            pathp[1].node = rbtn_right_get(pathp->node);
+            if (cmp == 0) {
+                /* Find node's successor, in preparation for swap. */
+                pathp->cmp = 1;
+                nodep = pathp;
+                for (pathp++; pathp->node; pathp++) {
+                    pathp->cmp = -1;
+                    pathp[1].node = rbtn_left_get(pathp->node);
+                }
+                break;
+            }
+        }
     }
-
-    /* If node's parent is black or node is root, back out. */
-    if (rb_color(rb_parent(node)) == RB_BLACK && rb_parent(node) != obj->head)
-        return;
-
-    /* Find out the identity */
-    map_node_t *parent = rb_parent(node), *grandparent = rb_parent(parent),
-               *uncle;
-
-    if (!rb_parent(parent))
-        return;
-
-    /* Find out the uncle */
-    if (grandparent->left == parent)
-        uncle = grandparent->right;
-    else
-        uncle = grandparent->left;
-
-    if (rb_is_red(uncle)) {
-        /* If the uncle is red, change color of parent and uncle to black */
-        rb_set_color(uncle, RB_BLACK);
-        rb_set_color(parent, RB_BLACK);
-
-        /* Change color of grandparent to red. */
-        rb_set_color(grandparent, RB_RED);
-
-        /* Call this on the grandparent */
-        map_fix_colors(obj, grandparent);
+    assert(nodep->node == node);
+    pathp--;
+    if (pathp->node != node) {
+        /* Swap node with its successor. */
+        bool tred = rbtn_red_get(pathp->node);
+        rbtn_color_set(pathp->node, rbtn_red_get(node));
+        rbtn_left_set(pathp->node, rbtn_left_get(node));
+        /* If node's successor is its right child, the following code */
```

Comment:
Use consistent comment style. That is,
```c
        /* If node's successor is its right child, the following code
         * will do the wrong thing for the right child pointer. 
         * However, it doesn't matter, because the pointer will be
         * properly set when the successor is pruned. 
         */
```